vpn/router.go

```diff
@@ -0,0 +1,110 @@
+package vpn
+
+import (
+	"net"
+	"net/netip"
+
+	"tailscale.com/wgengine/router"
+)
+
+func NewRouter(t *Tunnel) router.Router {
+	return &vpnRouter{tunnel: t}
+}
+
+type vpnRouter struct {
+	tunnel *Tunnel
+}
+
+func (*vpnRouter) Up() error {
+	// On macOS, this needs to bring the interface up: `ifconfig <interface> up`.
+	// On Windows, this is a no-op.
+	// How can we/Do we need to initiate this from within the dylib?
+	return nil
+}
+
+func (v *vpnRouter) Set(cfg *router.Config) error {
+	req := convertRouterConfig(cfg)
+	return v.tunnel.ApplyNetworkSettings(v.tunnel.ctx, req)
+}
+
+func (*vpnRouter) Close() error {
+	// There's no cleanup that we need to initiate from within the dylib.
+	return nil
+}
+
+func convertRouterConfig(cfg *router.Config) *NetworkSettingsRequest {
+	v4LocalAddrs := make([]string, 0)
+	v6LocalAddrs := make([]string, 0)
+	for _, addrs := range cfg.LocalAddrs {
+		if addrs.Addr().Is4() {
+			v4LocalAddrs = append(v4LocalAddrs, addrs.String())
+		} else if addrs.Addr().Is6() {
+			v6LocalAddrs = append(v6LocalAddrs, addrs.String())
+		} else {
+			continue
+		}
+	}
+	v4Routes := make([]*NetworkSettingsRequest_IPv4Settings_IPv4Route, 0)
+	v6Routes := make([]*NetworkSettingsRequest_IPv6Settings_IPv6Route, 0)
+	for _, route := range cfg.Routes {
+		if route.Addr().Is4() {
+			//nolint:gosec // outdated rule
```

Comment:
Interestingly, go 1.22 _does_ reuse the memory for `route` each time. I verified this with my debugger, that the pointer passed to the convert functions is the same memory address.

However, when I added some printf `%p` statements, it allocates new memory each time.

The go compiler is smart enough to realize that the convert functions don't actually hang on to the memory reference, so it can save a little space and avoid allocating on the heap.

Still, `netip.Prefix` is a regular struct, so it's probably better to pass by value rather than disabling _every_ gosec rule on this line.

Reply:
nice! I often forget that compilers can just optimize out (obviously) unnecessary memcpys.